dags/common/pull_ftp.py

```diff
@@ -6,28 +6,85 @@
 from airflow.api.common import trigger_dag
 from common.repository import IRepository
 from common.sftp_service import SFTPService
+from structlog import PrintLogger
 
 
-def migrate_from_ftp(sftp: SFTPService, repo: IRepository):
-    filenames = sftp.list_files()
+def migrate_files(filenames, sftp: SFTPService, repo: IRepository, logger: PrintLogger):
+    logger.msg(f"Processing files : {filenames}")
+    extracted_filenames = []
     for file in filenames:
+        logger.msg(f"Getting file {file} from SFTP.")
         file_bytes = sftp.get_file(file)
         if not zipfile.is_zipfile(file_bytes):
+            logger.msg("File is not a zipfile, processing next file.")
             continue
         with zipfile.ZipFile(file_bytes) as zip:
             for zip_filename in zip.namelist():
                 file_prefix = file.split(".")[0]
                 file_content = zip.read(zip_filename)
-                repo.save(file_prefix + "/" + zip_filename, io.BytesIO(file_content))
+                s3_filename = file_prefix + "/" + zip_filename
+                repo.save(s3_filename, io.BytesIO(file_content))
+                if repo.is_meta(s3_filename):
+                    extracted_filenames.append("extracted/" + s3_filename)
         repo.save(file, file_bytes)
+    return extracted_filenames
+
+
+def force_pull(sftp: SFTPService, repo: IRepository, logger: PrintLogger):
+    filenames = sftp.list_files()
+    return migrate_files(filenames, sftp, repo, logger)
+
+
+def migrate_from_ftp(
+    sftp: SFTPService, repo: IRepository, logger: PrintLogger, **kwargs
+):
+    filenames = []
+    params = kwargs["params"]
+    if "force_pull" in params and params["force_pull"]:
```

Comment:
I would split it in three different functions. Are the `params` passed from airflow? can we just have them as kwargs with defaults?

Reply:
Done. The params are passed by the user from the Airflow webview. I added defaults in the DAG 👍